spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/SparkParquetFileMergeRunner.java

```diff
@@ -0,0 +1,381 @@
+    // Check if all files are Parquet format
+    boolean allParquet =
+        group.rewrittenFiles().stream().allMatch(file -> file.format() == FileFormat.PARQUET);
+
+    if (!allParquet) {
+      LOG.debug("Cannot use row-group merge: not all files are Parquet format");
+      return false;
+    }
+
+    // Check if table has a sort order - row-group merge cannot preserve sort order
+    if (table().sortOrder().isSorted()) {
+      LOG.debug(
+          "Cannot use row-group merge: table has a sort order ({}). "
+              + "Row-group merging would not preserve the sort order.",
+          table().sortOrder());
+      return false;
+    }
+
+    // Check for delete files - row-group merge cannot apply deletes
+    boolean hasDeletes = group.fileScanTasks().stream().anyMatch(task -> !task.deletes().isEmpty());
+
+    if (hasDeletes) {
+      LOG.debug(
+          "Cannot use row-group merge: files have delete files or delete vectors. "
+              + "Row-group merging cannot apply deletes.");
+      return false;
+    }
+
+    // Validate schema compatibility using Iceberg InputFile API
+    try {
+      List<org.apache.iceberg.io.InputFile> inputFiles =
+          group.rewrittenFiles().stream()
+              .map(f -> table().io().newInputFile(f.path().toString()))
+              .collect(Collectors.toList());
+
+      boolean canMerge = ParquetFileMerger.canMerge(inputFiles);
+
+      if (!canMerge) {
+        LOG.warn(
+            "Cannot use row-group merge: schema validation failed for {} files. "
+                + "Falling back to standard rewrite.",
+            group.rewrittenFiles().size());
+        return false;
+      }
+
+      // TODO: Add encryption check here
+      // if (hasEncryptedFiles(group)) {
+      //   LOG.warn("Cannot use row-group merge: encrypted files detected");
+      //   return false;
+      // }
+
+      return true;
+    } catch (Exception e) {
+      LOG.warn("Cannot use row-group merge: validation failed", e);
+      return false;
+    }
+  }
+
+  /**
+   * Merges Parquet files in the group, respecting the target file size and expected output file
+   * count. Files are grouped into batches based on maxOutputFileSize, and each batch is merged into
+   * a separate output file.
+   */
+  private void mergeParquetFilesDistributed(String groupId, RewriteFileGroup group) {
+    PartitionSpec spec = table().specs().get(group.outputSpecId());
+    StructLike partition = group.info().partition();
+    long maxOutputFileSize = group.maxOutputFileSize();
+    int expectedOutputFiles = group.expectedOutputFiles();
+
+    LOG.info(
+        "Merging {} Parquet files into {} expected output files (group: {}, max size: {})",
+        group.rewrittenFiles().size(),
+        expectedOutputFiles,
+        groupId,
+        maxOutputFileSize);
+
+    // Group files into batches based on target size
+    List<List<DataFile>> fileBatches = groupFilesBySize(group.rewrittenFiles(), maxOutputFileSize);
```

Comment:
I still think that the no groupping should happen in the runner.
The planner should plan the correctly sized groups.